blockchain/chain_sync/src/network_context.rs

```diff
@@ -62,6 +65,47 @@ impl<DB: Clone> Clone for SyncNetworkContext<DB> {
     }
 }
 
+type RaceBatchFuture<T> = Pin<Box<dyn Future<Output = Result<T, String>> + Send>>;
+
+struct RaceBatch<T> {
+    tasks: JoinSet<Result<T, String>>,
+    semaphore: Arc<Semaphore>,
+}
+
+impl<T> RaceBatch<T>
+where
+    T: Send + 'static,
+{
+    pub fn new(max_concurrent_jobs: usize) -> Self {
+        RaceBatch {
+            tasks: JoinSet::new(),
+            semaphore: Arc::new(Semaphore::new(max_concurrent_jobs)),
+        }
+    }
+
+    pub fn add(&mut self, future: RaceBatchFuture<T>) {
```

Comment:
I don't think the `Pin` is necessary. Something like this should also work:
```suggestion
    pub fn add(&mut self, future: impl Future<Output = Result<T, String>> + Send + 'static) {
```